consensus-types/hdiff/state_diff.go

```diff
@@ -0,0 +1,2040 @@
+	tPendingConsolidations, err := target.PendingConsolidations()
+	if err != nil {
+		return err
+	}
+	tlen := len(tPendingConsolidations)
+	tPendingConsolidations = append(tPendingConsolidations, nil)
+	var sPendingConsolidations []*ethpb.PendingConsolidation
+	if source.Version() >= version.Electra {
+		sPendingConsolidations, err = source.PendingConsolidations()
+		if err != nil {
+			return err
+		}
+	}
+	tPendingConsolidations = append(tPendingConsolidations, sPendingConsolidations...)
+	index := kmpIndex(len(sPendingConsolidations), tPendingConsolidations, helpers.PendingConsolidationsEqual)
+	diff.pendingConsolidationsIndex = uint64(index)
+	diff.pendingConsolidationsDiffs = make([]*ethpb.PendingConsolidation, tlen+index-len(sPendingConsolidations))
+	for i, d := range tPendingConsolidations[len(sPendingConsolidations)-index : tlen] {
+		diff.pendingConsolidationsDiffs[i] = &ethpb.PendingConsolidation{
+			SourceIndex: d.SourceIndex,
+			TargetIndex: d.TargetIndex,
+		}
+	}
+	return nil
+}
+
+// applyValidatorDiff applies the validator diff to the source state in place.
+func applyValidatorDiff(source state.BeaconState, diff []validatorDiff) (state.BeaconState, error) {
+	sVals := source.Validators()
+	if len(sVals) < len(diff) {
+		return nil, errors.Errorf("target validators length %d is less than source %d", len(diff), len(sVals))
+	}
+	for _, d := range diff {
+		if d.index > uint32(len(sVals)) {
+			return nil, errors.Errorf("validator index %d is greater than length %d", d.index, len(sVals))
+		}
+		if d.index == uint32(len(sVals)) {
+			// A valid diff should never have an index greater than the length of the source validators.
+			sVals = append(sVals, &ethpb.Validator{})
+		}
+		if d.PublicKey != nil {
+			sVals[d.index].PublicKey = slices.Clone(d.PublicKey)
+		}
+		if d.WithdrawalCredentials != nil {
+			sVals[d.index].WithdrawalCredentials = slices.Clone(d.WithdrawalCredentials)
+		}
+		sVals[d.index].EffectiveBalance = d.EffectiveBalance
+		sVals[d.index].Slashed = d.Slashed
+		sVals[d.index].ActivationEligibilityEpoch = d.ActivationEligibilityEpoch
+		sVals[d.index].ActivationEpoch = d.ActivationEpoch
+		sVals[d.index].ExitEpoch = d.ExitEpoch
+		sVals[d.index].WithdrawableEpoch = d.WithdrawableEpoch
+	}
+	if err := source.SetValidators(sVals); err != nil {
+		return nil, errors.Wrap(err, "failed to set validators")
+	}
+	return source, nil
+}
+
+// applyBalancesDiff applies the balances diff to the source state in place.
+func applyBalancesDiff(source state.BeaconState, diff []int64) (state.BeaconState, error) {
+	sBalances := source.Balances()
+	if len(diff) < len(sBalances) {
+		return nil, errors.Errorf("target balances length %d is less than source %d", len(diff), len(sBalances))
+	}
+	sBalances = append(sBalances, make([]uint64, len(diff)-len(sBalances))...)
+	for i, t := range diff {
+		if t >= 0 {
+			sBalances[i] += uint64(t)
+		} else {
+			sBalances[i] -= uint64(-t)
```

Comment:
do you want to guard against underflow here?

Reply:
nope, this applies a diff, the resulting state has to be a valid state.

Reply:
what if the source state is different than the diff 's source state

Reply:
apply Balances will return an invalid result in this case.